### ISSUE TITLE
September Iteration Plan

### ISSUE THREAD

**Issue by kieferrm** [Microsoft/vscode#11917]
## Endgame Schedule
- *September, 26th* Code freeze for the endgame
- *September, 30th* Endgame done

##### Verification
- [ ] Verification ([wiki](https://github.com/Microsoft/vscode/wiki/Issue-Tracking#verification)): since there are quite some old bugs to verify we verify by age, youngest first. The queries are:
    - [x] [Verification needed](https://github.com/Microsoft/vscode/issues?q=is%3Aissue+-label%3Averified+is%3Aclosed+label%3Averification-needed)
    - [ ] [September Bugs](https://github.com/Microsoft/vscode/issues?page=1&q=is%3Aissue+label%3Abug+-label%3Averified+is%3Aclosed+-label%3Aduplicate+-label%3Ainvalid+closed%3A%3E%3D2016-09-01&utf8=%E2%9C%93)

##### Monday
- [x] Code freeze at 5pm PT
- [x] Ensure we have a green build on all platforms
- [x] All test items contain sufficiently comprehensive test descriptions by 6pm PT

##### Tuesday
- [x] Test build starts at 7am CET / 10pm PT on Monday
- [x] Test plan ready by 8am CET / 11pm PT on Monday
- [x] Testing

##### Wednesday
- [x] Testing
- [x] Remind team members to assign issues that they intend to fix to the September milestone
- [x] Fixing (self-assigned, milestone assigned)
- [x] Verification (queries above)

##### Thursday
- [x] Fixing (scrutiny sets in - major bugs only - to be discussed in stand-up meeting, labeled as `candidate`)
- [x] Verification (queries above)
- [x] Add/update shrink-wrap files for built-in extensions if needed (see [instructions](https://github.com/Microsoft/vscode/issues/8570#issuecomment-229669456)) (was done along the way)
- [x] Update `OSSREADME.json` for built-in extensions based on differences to generated `npm-shrinkwrap.json` files if needed **@alexandrudima**/**dbaeumer**
- [x] Run [OSS tool](https://github.com/Microsoft/vscode-distro/blob/master/distro-tools/README.md) after merging shrink-wrap findings **@alexandrudima**/**dbaeumer**
  - *The LCA review of the ThirdPartyNotices.txt files is not needed anymore*
- [x] Check new OSS usage is entered into the [OSS registry](https://ossmsft.visualstudio.com/_apps/hub/ms.vss-oss-web.hub-oss) **@alexandrudima**/**dbaeumer**

##### Friday
- Satellite modules/npm packages ready, version updated, smoke tested
  - [x] vscode **@bpasero**
  - [x] yo generator **@bpasero**
  - [x] vsce **@joaomoreno**
  - [x] node debug **@isidorn**
- [x] Translation input - **@dbaeumer**
- [x] Verification (queries above)
- [x] Fixing (only critical bugs - no string changes)
- [Smoketest](https://github.com/Microsoft/vscode/wiki/Smoke-Test)
  - [x] Windows - **@chrmarti**
  - [x] OS X - **@ramya-rao-a**
  - [x] Linux - **@mjbvz**
- [x] All release notes updated. Release notes are collected in a file named *`1_6.md`* in this [repo directory](https://github.com/Microsoft/vscode-docs/blob/vnext/release-notes/)
    - [x] @alexandrudima 
    - [x] @weinand 
    - [x] @bpasero 
    - [x] @chrmarti 
    - [x] @tyriar 
    - [x] @dbaeumer
    - [x] @egamma 
    - [x] @isidorn 
    - [x] @joaomoreno 
    - [x] @jrieken 
    - [x] @kieferrm 
    - [x] @aeschli 
    - [x] @rebornix 
    - [x] @octref 
    - [x] @ramya-rao-a 
    - [x] @roblourens 
    - [x] @sandy081
- [x] Acknowledge pull requests in release notes. We acknowledge PRs from outside the team. **@all**
    - [x] @alexandrudima 
    - [x] @weinand 
    - [x] @bpasero 
    - [x] @chrmarti 
    - [x] @tyriar 
    - [x] @dbaeumer
    - [x] @egamma 
    - [x] @isidorn 
    - [x] @joaomoreno 
    - [x] @jrieken 
    - [x] @kieferrm 
    - [x] @aeschli 
    - [x] @rebornix 
    - [x] @octref 
    - [x] @ramya-rao-a 
    - [x] @roblourens 
    - [x] @sandy081
- [x] Mention notable fixes in the release notes:
    - [x] @alexandrudima 
    - [x] @weinand 
    - [x] @bpasero 
    - [x] @chrmarti 
    - [x] @tyriar 
    - [x] @dbaeumer
    - [x] @egamma 
    - [x] @isidorn 
    - [x] @joaomoreno 
    - [x] @jrieken 
    - [x] @kieferrm 
    - [x] @aeschli 
    - [x] @rebornix 
    - [x] @octref 
    - [x] @ramya-rao-a 
    - [x] @roblourens 
    - [x] @sandy081

##### Friday/Monday
- When done fixing/verifying and there are changes since last build at the end of day PT
  - [x] Trigger new insider build and publish it manually **@chrmarti**
- [x] Pause scheduled `insider` builds **@chrmarti**
- [x] Branch code to `release/1.6` **@chrmarti**
- [x] Announce master is open for business **@chrmarti**
- [x] Polish release notes **@gregvanl**

##### Wednesday
- [x] Publish `Insider` with hand-picked and reviewed candidate fixes **@chrmarti**

##### Thursday
- [x] Publish `Insider` with hand-picked and reviewed candidate fixes **@chrmarti**

##### Monday
- [x] Build stable for all platforms **@chrmarti**
- [x] Sanity check of installable bits
  - [x] Windows
    - [x] signed installer **@chrmarti**
    - [x] zip **@chrmarti**
  - [x] OS X **@chrmarti**
  - [x] Linux
    - [x] deb package 64-bit **@tyriar**
    - [x] deb package 32-bit **@tyriar**
    - [x] rpm package 64-bit **@chrmarti**
    - [x] rpm package 32-bit **@chrmarti**
    - [x] archives **@tyriar**
- [x] Publish website **@gregvanl**
- [x] Release [stable](https://vscode-builds.azurewebsites.net/stable) (commit: e52fb0bc) **@chrmarti**
- [x] Add a git tag to `HEAD` of `release/1.6` in format `1.6.z` **@chrmarti**
- [x] Twitter announcement **@chrisdias**
- [x] Enable scheduled `insider` builds **@chrmarti**

## General

This plan captures what we work on during September. We are back to our normal 4 week rhythm. We will ship **early October**.

Not unusual and still noteworthy: We will do several explorations that might or might not result in concrete deliverables in this or later releases.

## Dates
- *September, 26th* Code freeze for the endgame
- *September, 30th* Endgame done

## Plan Items

| Mark  | Description |
| ------------- | ------------- |
| :runner: | work in progress  |
| :hand: | blocked task |
| :muscle: | stretch goal for this iteration  |
| :large_blue_circle: | more details required |
| :red_circle: | missing issue reference |

### Feedback

Issues to address community feedback:

- [ ] :muscle: Workbench - Support to split editors horizontally #1749 **@bpasero**
- [ ] :runner: :muscle: Workbench - Preserve unsaved files on quit #101 **@tyriar**
- [x] Workbench - Support to show icons in more places (tabs, open editors, quick open) #11751 **@bpasero**
  - Test #12525
  - [x] Release notes
- [x] Find/Search - Support history for find/search patterns #5508 **@sandy081**
  - Test #12593
  - [ ] Release notes
- [x] Extensions - support to recommend extensions for a workspace #11936 **@sandy081**
  - Test #12594
  - [ ] Release notes
- [ ] :runner: Extensions - support to disable an extension #2882 **@joaomoreno** **@sandy081**
- [x] :runner: Extensions - support to publish and install a set of extensions #11937 **@joaomoreno** **@sandy081**
  - Test #12647
  - [ ] Release notes
- [x] API - Debug: Support launch.json participation from an extension #9061 (depends on #1396, #8937) **@isidorn**
  - Test #12460
  - [x] Release notes
- [x] API - enable format on save scenario, requires sync onWillSaveDocument hook #239, #12449 **@jrieken**
  - Test #12640 #12639
  - [ ] Release notes
- [x] API - Action terminal API feedback #10917 #10925 #11275 #11384 #11919 **@tyriar**
  - Test #12595
  - [ ] Release notes
- [ ] :runner: Debug - Improve Docker/SourceMaps scenarios #6915, #9742 **@weinand**
- [x] Debug - add globbing support for the `outdir` attribute in the launch config #6915 **@weinand**
  - Test #12254
  - [x] Release notes
- [x] Debug - Add launch option that allows run node project by npm script #2726 **@weinand**
  - Test #12253
  - [x] Release notes
- [x] Search/quick open - Improve quick open performance on windows (some improvements got backed out)  #11182 **@chrmarti**
- :hand: Install - Support Linux auto update (signing support should be in place now) #229 #2973 **@tyriar**

### Debt Items

Issues to reduce our engineering debt:

- [ ] :runner: Revisit and improve the use of the editor inside the debug console **@alexandrudima** #12020
- [x] Extract HTML language support into an extension  #8938 **@aeschli**
  - Test #12100
  - [ ] Release notes
- [x] Find a better way to cache objects between main and ext host #11455 **@jrieken**
  - [ ] Release notes
- [x] :runner: Markdown - tmGrammar needs to be improved - [see query](https://github.com/Microsoft/vscode/labels/markdown) **@mjbvz**
  - Test #12652
  - [ ] Release notes
- [x] Publish source maps for built-in extensions #10994 **@joaomoreno**
  - Test #12597
  - [x] Release notes
- [x] Rendering of documentation in Intellisense and hover is inconsistent #2330 **@joaomoreno**
  - Test #12598
  - [x] Release notes
- [x] Search/Problems - More consistent highlighting using range highlight (find, search, problems, outline, go to line) #11318 #8365 **@sandy081**
  - Test #12599
  - [ ] Release notes

### Extension Contributions

Support/contribute to extensions:

- [x] :runner: Extension - Handle the case that an extension is for a newer VS Code version, enable to publish extension that require new API (e.g. insiders only) gracefully #11573 **@joaomoreno**, **@sandy081**
  - Test #12658
  - [x] Release Notes
- [x] :runner: Chrome debugger - align the functionality of the node-debugger and chrome debugger Microsoft/vscode-node-cdp-debug#7  **@roblourens**
  - Test #12600
  - [x] Release Notes
- [ ] :runner: VIM extension support see #8997 **@rebornix**
- [x] :runner: VIM - setting for relative line numbers directly in Code #9561 **@rebornix**
  - Test #12653
  - [x] Release Notes
- [x] :runner: Go - support ongoing maintenance **@ramya-rao-a**

       1.  Fix installation when goimports is the chosen formatting tool
       2.  Use diff output from gorename to allow undo on rename operation
       3.  Add configurable test flags to run Go tests
       4.  Replace full vendor paths with relative paths in Add Import command
       5.  Add autocomplete feature for methods from unimported packages
       6.  Allow configurable env vars to be used while running Go tests

- [x] vscode-eslint - assist user in setting up eslint for the workspace **@dbaeumer**
  - Test #12601
  - [ ] Release notes
- [x] vsocde-eslint - don't lint when there are config errors, support to disable eslint **@dbaeumer**
  - Test #12602
  - [ ] Release notes
- [x] vscode-tslint - align vscode-tslint with the vscode-eslint improvements Microsoft/vscode-tslint#86 **@egamma**
  - Test #12603
  - [ ] Release notes
- [x] vscode-tslint - Improve and document the integration of vscode-tslint with a gulp tslint task. #89 **@egamma**
  - Test #12604
  - [ ] Release notes

### Engineering

This section lists several engineering/development items that the team identified:

- [x] Electron - investigate into adoption of latest version #10206 **@bpasero**
  - Test #12519
  - [x] Release notes
- :hand: Explore open/install an extension from the Market Place #10340 **@joaomoreno**
- [x] Show release notes in product #10179 **@joaomoreno**
  - Test #12520
  - [x] Release notes
- [x] Inform users that they are running a patched version of VS Code #12021 **@alexandrudima**
  - Test #12659
  - [x] Release notes
- [x] Build - Support bundling extensions from Marketplace at build time  #12018 @joaomoreno
  - Test #12521
  - [x] Release notes
- [x] Debugger Protocol - Breakpoint hit counts Microsoft/vscode-debugadapter-node#37 **@weinand**
  - [x] Release notes
- [x] Debugger Protocol - Provide a JSON schema for the debug protocol Microsoft/vscode-debugadapter-node#51 **@weinand**
  - [x] Release notes
- [x] Debugger Protocol - External Terminal support for runInTerminal #11581 **@weinand**
  - Test #12252
  - [x] Release notes
- [x] Debug/tasks - Consolidate variable substitution between debug and tasks #8754 **@isidorn**
  - Test #12384
  - [x] Release notes
- [x] Settings - enable programmatic changes to settings (includes launch.json, task.json) #8937 **@bpasero**
  - Test #12460
  - [x] Release notes
- [x] Settings -  expose API to change a user settings, launch.json, task.json #1396 **@jrieken**
  - Test #12527 #12614
  - [ ] Release notes
- [x] Settings - API adopt new settings API (file associations, enable auto save from the file menu) #11974 **@bpasero**
  - Test #12527
  - [x] Release notes
- [x] TypeScript - Adopt TS 2.0 and bundle 2.0 for September #10136 **@dbaeumer**
  - Test #12633
  - [x] Release notes
- [x] TypeScript - Inform user about TS 2.0 in the product  (remove the tsdk guidance) #12091 **@dbaeumer**
  - Test #12633
  - [x] Release notes
- [ ] :muscle: Language server - shutdown the server when idle for an extended period Microsoft/vscode-languageserver-node#97 **@dbaeumer**
- [ ] :muscle: Language server - throttle validation when document changes Microsoft/vscode-languageserver-node#42 **@dbaeumer**

### Explorations

We are continuously exploring options and ideas. Some might make it into future releases.

- [ ] Investigate into using SCCS instead of CSS in VS Code #8589 **@bpasero** **@aeschli** **@tyriar**
- [ ] Workbench - Support for running tests, see results, navigate to failures #9505 **@alexandrudima**
- [x] :runner: :large_blue_circle: Initial configuration on first run (implementation depends on settings API) #12534 **@chrmarti**
  - Test #12607
  - [x] Release notes
- [x] Node - investigate into 'configure workspace' action for node **@jrieken**
- [x] Debug - investigate in presenting inline values #12256 @weinand
- [ ] Debug Console: investigate into replacing the tree widget with a text editor #11462 **@isidorn**
- [ ] :runner: :muscle: API - explore inline UI extensions (peak, split editor) #3220 **@jrieken**, **@DonJayamanne**
- [ ] :runner: :muscle: API - declarative contribution of custom explorers (based on tree models) **@tyriar** **@octref**
- [x] 'Open in VS Code' action from Github  **@joaomoreno**
- [ ] :runner: TypeScript - investigate automatic type acquisition provided by the tsserver #12458 **@kieferrm** 
- [ ] PHP - investigate how and where to contribute **@mousetraps**
- [ ] :large_blue_circle: :muscle: - Serviceability improvements when VS Code crashes **@ramya-rao-a**

### Feedback Channels

- [x] :runner: Telemetry - Track Autocomplete suggestions and terminal events in xflow and Targets report **@mjbvz**
- [x] Telemetry - Improve data refresh times by improving SQL queries in Extensions, Targets and LongTermUse reports in Power BI **@ramya-rao-a** **@mjbvz**
- [x] Telemetry - Explore BigQuery as a source instead of SQL **@mjbvz**

### UX
- [ ] :runner: Manage panels on-going. Prereq for maximizing a panel **@isidorn** **@stevencl** **@bgashler1**
- [ ] :muscle: :runner: Advanced refactoring UI **@stevencl** **@bgashler1**

**Comment by felixfbecker** [comment]
>  PHP - investigate how and where to contribute @mousetraps 

https://github.com/felixfbecker/vscode-php-intellisense
https://github.com/felixfbecker/php-language-server

I would love to get more contributions, already got some from the Eclipse Che folks 🙂

**Comment by csholmq** [comment]
https://github.com/Microsoft/vscode-cpptools isn't moving very fast apart from debugging features. IntelliSense is largely missing. Perhaps it could use a team boost?

**Comment by glen-84** [comment]
> :red_circle: Language server - throttle validation when document changes @dbaeumer

This is https://github.com/Microsoft/vscode-languageserver-node/issues/42, if I'm not mistaken.

**Comment by egamma** [comment]
@glen-84 thanks, I've updated the item.

**Comment by nevadascout** [comment]
> PHP - investigate how and where to contribute @mousetraps

If you're looking for a pure javascript implementation of a php language server, my project **https://github.com/HvyIndustries/crane** is progressing nicely.

I'm currently adding go to definition functionality and optimising/improving stability

**Comment by egamma** [comment]
Closing VS Code 1.6 has been pushed to stable.

**Comment by egamma** [comment]
@daviwil fyi,  moving the milestone back to September. We only assign candidate bugs for a recovery build to the recovery milestone.

**Comment by daviwil** [comment]
Wow... I have no idea how I did that! Certainly wasn't on purpose. Sorry about that!
